src/main/java/org/manifold/compiler/front/MappedArray.java

```diff
@@ -0,0 +1,163 @@
+package org.manifold.compiler.front;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+import java.util.*;
+
+public class MappedArray<K, V> implements Iterable<MappedArray<K, V>.Entry> {
+
+  private List<V> data;
+  private BiMap<K, Integer> keyedIndices = HashBiMap.create();
+
+  public MappedArray(Map<K, V> data) {
+    this.data = new ArrayList<>(data.size());
+
+    int i = 0;
+    for (Map.Entry<K, V> entry : data.entrySet()) {
+      this.data.add(entry.getValue());
+      keyedIndices.put(entry.getKey(), i);
+    }
+  }
+
+  public MappedArray(List<V> data) {
+    this.data = data;
```

Comment:
Yes, a deep copy would probably help us avoid issues with shared data in the future. I anticipate that this could lead to some nasty bugs. There's also the performance consideration of allowing the user to provide some arbitrary `List` implementation here which might not be the optimal one for this data structure. So, I'd prefer a deep copy if that's okay -- I understand the performance issue involved in doing this operation but I think it's acceptable in order to make this a bit safer for general use (especially since it is *not* documented right now :grin:)